sysboundary/ionosphere.cpp

```diff
@@ -774,15 +3685,128 @@ namespace SBC {
+         const Real zmin = cellParams[CellParams::ZCRD];
+         const Real xmax = xmin + cellParams[CellParams::DX];
+         const Real ymax = ymin + cellParams[CellParams::DY];
+         const Real zmax = zmin + cellParams[CellParams::DZ];
+         const Real xcen = 0.5*(xmin+xmax);
+         const Real ycen = 0.5*(ymin+ymax);
+         const Real zcen = 0.5*(zmin+zmax);
+         std::array< std::array<Real, 3>, 6> tracepoints;
+         tracepoints[0] = {xmin, ycen, zcen};
+         tracepoints[1] = {xmax, ycen, zcen};
+         tracepoints[2] = {xcen, ymin, zcen};
+         tracepoints[3] = {xcen, ymax, zcen};
+         tracepoints[4] = {xcen, ycen, zmin};
+         tracepoints[5] = {xcen, ycen, zmax};
+         std::array<Real, 6> potentials;
+
+         for(int i=0; i<6; i++) {
+            // Get potential at each of these 6 points
+            potentials[i] = ionosphereGrid.interpolateUpmappedPotential(tracepoints[i]);
+         }
+
+         // Calculate E from potential differences as E = -grad(phi)
+         Vec3d E({
+               (potentials[0] - potentials[1]) / cellParams[CellParams::DX],
+               (potentials[2] - potentials[3]) / cellParams[CellParams::DY],
+               (potentials[4] - potentials[5]) / cellParams[CellParams::DZ]});
+         Vec3d B({
+               cellParams[CellParams::BGBXVOL],
+               cellParams[CellParams::BGBYVOL],
+               cellParams[CellParams::BGBZVOL]});
+
+         // Add E from neutral wind convection for all cells with L <= 5
+         Vec3d Omega(0,0,Ionosphere::earthAngularVelocity); // Earth rotation vector
+         Vec3d r(xcen,ycen,zcen);
+         Vec3d vn = cross_product(Omega,r);
+
+         Real radius = vector_length(r);
+         if(radius/physicalconstants::R_E <= Ionosphere::plasmapauseL * (r[0]*r[0] + r[1]*r[1]) / (radius*radius)) {
+            E -= cross_product(vn, B);
+         }
+
+         const Real Bsqr = B[0]*B[0] + B[1]*B[1] + B[2]*B[2];
+
+         // Calculate cell bulk velocity as E x B / B^2
+         std::array<Real, 3> vDrift;
+         vDrift[0] = (E[1] * B[2] - E[2] * B[1])/Bsqr;
+         vDrift[1] = (E[2] * B[0] - E[0] * B[2])/Bsqr;
+         vDrift[2] = (E[0] * B[1] - E[1] * B[0])/Bsqr;
+
+         // Average density and temperature from the nearest cells
+         const vector<CellID> closestCells = getAllClosestNonsysboundaryCells(cellID);
+         Real density = 0, pressure = 0;
+         for (CellID celli : closestCells) {
+            density += mpiGrid[celli]->parameters[CellParams::RHOM];
+            pressure += mpiGrid[celli]->parameters[CellParams::P_11] + mpiGrid[celli]->parameters[CellParams::P_22] + mpiGrid[celli]->parameters[CellParams::P_33];
+         }
+         density /= closestCells.size()*physicalconstants::MASS_PROTON;
+         pressure /= 3.0*closestCells.size();
+         // TODO make this multipop
+         creal temperature = pressure / (density * physicalconstants::K_B);
```

Comment:
This bit should be changed still before merging.

Reply:
Yup, done. This has now been replaced by the ```innerBoundaryVDFmode``` parameter, which has the possible values "FixedMoments", "AverageMoments", "AverageAllMoments" and "CopyAndLosscone". Default value is FixedMoments, as that seems to have worked best.